core/pta/rockchip/rk_secure_boot.c

```diff
@@ -0,0 +1,406 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2025, Pengutronix, Michael Tretter <entwicklung@pengutronix.de>
+ */
+
+#include <config.h>
+#include <drivers/rockchip_otp.h>
+#include <kernel/pseudo_ta.h>
+#include <kernel/tee_misc.h>
+#include <tee/uuid.h>
+#include <utee_defines.h>
+#include <stdio.h>
+#include <string.h>
+#include <platform_config.h>
+
+#include <pta_rk_secure_boot.h>
+
+#define PTA_NAME "rk_secure_boot.pta"
+
+/*
+ * Disable simulation to actually fusing the hash into the OTP. Enabling this
+ * option is necessary to actually enable secure boot with PTA, but may
+ * potentially brick your device.
+ */
+static const int simulation = IS_ENABLED2(CFG_RK_SECURE_BOOT_SIMULATION);
+
+static inline bool bit_test(uint32_t value, uint32_t bit)
+{
+	return (value & bit) == bit;
+}
+
+/* Convert hash from 32 bit words in OTP into a byte array */
+static TEE_Result otp_to_bytes(uint32_t *otp, size_t otp_size,
+			       uint8_t *bytes, size_t bytes_size)
+{
+	uint32_t *u32 = otp;
+	uint8_t *u8 = bytes;
+	size_t i;
+	size_t j;
+
+	if (bytes_size * sizeof(*bytes) != otp_size * sizeof(*otp))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (i = 0; i < otp_size; i++, u32++)
+		for (j = 0; j < sizeof(*u32); j++, u8++)
+			*u8 = (uint8_t)(*u32 >> (j * 8));
+
+	return TEE_SUCCESS;
+}
+
+/* Convert hash from byte array into 32 bit words for OTP */
+static TEE_Result bytes_to_otp(uint8_t *bytes, size_t bytes_size,
+			       uint32_t *otp, size_t otp_size)
+{
+	uint32_t *u32 = otp;
+	uint8_t *u8 = bytes;
+	size_t i;
+	size_t j;
+
+	if (bytes_size * sizeof(*bytes) != otp_size * sizeof(*otp))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (i = 0; i < otp_size; i++, u32++) {
+		*u32 = 0;
+		for (j = 0; j < sizeof(*u32); j++, u8++)
+			*u32 += (uint32_t)*u8 << (j * 8);
+	}
+
+	return TEE_SUCCESS;
+}
+
+#define HASH_STRING_SIZE 88
+static TEE_Result otp_to_string(uint32_t *otp, size_t otp_size,
+				char *str, size_t str_size)
+{
+	if (otp_size != 8 || str_size != HASH_STRING_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	snprintf(str, str_size,
+		 "0x%08x 0x%08x 0x%08x 0x%08x 0x%08x 0x%08x 0x%08x 0x%08x",
```

Comment:
Please use `PRIx32` instead of "x".

Reply:
I changed all places that used `%x` to `PRIx32`.